velox/expression/CastExpr-inl.h

```diff
@@ -296,6 +296,52 @@ VectorPtr CastExpr::applyDecimalToBooleanCast(
   return result;
 }
 
+template <typename FromNativeType>
+VectorPtr CastExpr::applyDecimalToVarcharCast(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& fromType) {
+  VectorPtr result;
+  context.ensureWritable(rows, VARCHAR(), result);
+  (*result).clearNulls(rows);
+  const auto simpleInput = input.as<SimpleVector<FromNativeType>>();
+
+  // Calculate the total size of the raw string buffer.
+  size_t totalResultBytes = 0;
+  applyToSelectedNoThrowLocal(context, rows, result, [&](int row) {
+    if (!simpleInput->isNullAt(row)) {
+      std::string output =
+          DecimalUtil::toString((int128_t)simpleInput->valueAt(row), fromType);
```

Comment:
DecimalUtil::toString looks quite inefficient and I believe not supposed to be used in hot paths. It would be nice to add a benchmark. See velox/benchmarks/basic/CastBenchmark.cpp

We may need to write a different version of DecimalUtil::toString that doesn't create intermediate strings. It would take a pointer to pre-allocated buffer and write directly into it. A decimal number has a precision and scale, hence, we know the max size of the serialized string and therefore can pre-allocate sufficient buffer.

Reply:
Understood. Working on the improvement. Thank you.

Reply:
@mbasmanova Just pushed the improved code and below are the benchmark results.

The previous `DecimalUtil::toString` way:
```
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
cast##try_cast_invalid_empty_input                        988.32us     1.01K
cast##tryexpr_cast_invalid_empty_input                    293.34ms      3.41
cast##try_cast_invalid_nan                                834.97ms      1.20
cast##tryexpr_cast_invalid_nan                               1.30s   766.33m
cast##try_cast_valid                                        6.43ms    155.62
cast##tryexpr_cast_valid                                    6.43ms    155.60
cast##cast_valid                                            6.44ms    155.40
cast##cast_short_decimal                                   45.43ms     22.01
cast##cast_long_decimal                                   132.09ms      7.57
```
Current impl.:
```
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
cast##try_cast_invalid_empty_input                          1.09ms    921.00
cast##tryexpr_cast_invalid_empty_input                    292.68ms      3.42
cast##try_cast_invalid_nan                                833.48ms      1.20
cast##tryexpr_cast_invalid_nan                               1.29s   777.43m
cast##try_cast_valid                                        6.42ms    155.85
cast##tryexpr_cast_valid                                    6.45ms    155.00
cast##cast_valid                                            6.42ms    155.75
cast##cast_short_decimal                                   18.34ms     54.52
cast##cast_long_decimal                                    63.35ms     15.79
```

Reply:
@rui-mo Thank you for benchmarking. How should I interpret the results? Any chance you could summarize these? Should I only compare cast##cast_short_decimal  and cast##cast_long_decimal ? In this case, the difference is ~2x?

Reply:
@mbasmanova Summarized  the benchmark results in above table. The two benchmarks are `cast##cast_short_decimal` and `cast##cast_long_decimal`, and this table shows the time for different versions of implementations. We should be noticed the benchmark does not compare the performance of the three functions `DecimalUtil::toString`, `std::to_string`, `std::to_chars`, but the whole implementation of cast from decimal to varchar. For example, the way to calculate the total buffer size varies in each implementation.

Reply:
@rui-mo Thank you for benchmarking. Is it fair to say that final implementation is 4-5x faster than the original one? It would be great to update PR description to explain the details of the final implementation and how it differs performance-wise from an original one.

Reply:
Yes, that's right. Updated the PR description.